clients/samples/example_bfdot_hip_bfloat16.cpp

```diff
@@ -0,0 +1,133 @@
+/* ************************************************************************
+ * Copyright (C) 2016-2023 Advanced Micro Devices, Inc. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * ************************************************************************ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <vector>
+
+#include "utility.h"
+#include <hipblas/hipblas.h>
+
+/* ============================================================================================ */
+
+// if just using host compiler, hip gives minimum definition of bfloat16, so doing
+// most of the work in fp32
+void reference_bfdot(int                             n,
+                     std::vector<hip_bfloat16>       hx,
+                     int                             incx,
+                     const std::vector<hip_bfloat16> hy,
+                     int                             incy,
+                     hip_bfloat16*                   res)
+{
+    float tmp = 0;
+    for(int i = 0; i < n; i++)
+        tmp += bfloat16_to_float(hx[i * size_t(incx)]) * bfloat16_to_float(hy[i * size_t(incy)]);
+    *res = float_to_bfloat16(tmp);
+}
+
+int main()
+{
+    int             N    = 1024;
+    int             incx = 1;
+    int             incy = 1;
+    hipblasStatus_t status;
+
+    // Naming: dX is in GPU (device) memory. hK is in CPU (host) memory, plz follow this practice
+    std::vector<hip_bfloat16> hx(N);
```

Comment:
Can you consider adding here
#ifdef HIPBLAS_USE_HIP_BFLOAT16
    //    the hip_bfloat16 usage
#else
    // the hipblasBfloat16 usage
#endif
This would be similar to https://github.com/ROCmSoftwarePlatform/rocBLAS/blob/release/rocm-rel-5.6/clients/samples/example_strmm_v3.cpp#L174 . It would enable the example to be built with or without HIPBLAS_USE_HIP_BFLOAT16 defined. It would also indicate what needs to change when using HIPBLAS_USE_HIP_BFLOAT16 .